crates/rbuilder/src/backtest/fetch/csv.rs

```diff
@@ -0,0 +1,113 @@
+pub struct CSVDatasource {
+    batches: HashMap<u64, Vec<TransactionSignedEcRecovered>>,
+}
+
+impl CSVDatasource {
+    pub fn new(filename: impl Into<PathBuf>) -> eyre::Result<Self> {
+        let batches = Self::load_transactions_from_csv(filename.into())?;
+        Ok(Self { batches })
+    }
+
+    fn load_transactions_from_csv(
+        filename: PathBuf,
+    ) -> eyre::Result<HashMap<u64, Vec<TransactionSignedEcRecovered>>> {
+        let file = File::open(&filename)
+            .wrap_err_with(|| format!("Failed to open file: {}", filename.display()))?;
+        let mut reader = Reader::from_reader(file);
+        let mut batches: HashMap<u64, Vec<TransactionSignedEcRecovered>> = HashMap::new();
+
+        for result in reader.records() {
+            let record = result?;
+            if record.len() != 2 {
+                return Err(eyre::eyre!("Invalid CSV format"));
+            }
+
+            let batch_number: u64 = record[0].parse()?;
+            let rlp_hex = &record[1];
+            let rlp_bytes = hex::decode(rlp_hex)?;
+            let tx = TransactionSignedEcRecovered::decode(&mut &rlp_bytes[..])?;
+
+            batches.entry(batch_number % 10).or_default().push(tx);
+        }
+
+        Ok(batches)
+    }
+}
+
+#[async_trait]
+impl DataSource for CSVDatasource {
+    async fn get_orders(&self, block: BlockRef) -> eyre::Result<Vec<OrdersWithTimestamp>> {
+        // The csv datasource is one with 10 batches, where batch is a list of transactions
+        // Since we don't have full "real" blocks, we'll just use the block number to determine the batch
+        // Thus the usage of mod 10 is just to determine the batch number that we get transactions from, e.g. block 100 corresponds to 0, 101 to 1, 109 to 9, etc.
+        let batch_number = block.block_number % 10;
+        let transactions = self.batches.get(&batch_number).cloned().unwrap_or_default();
+
+        let mut uuid_num = 0;
+        let orders: Vec<OrdersWithTimestamp> = transactions
+            .into_iter()
+            .map(|tx| {
+                let order = transaction_to_order(block.block_number, &mut uuid_num, tx);
+                OrdersWithTimestamp {
+                    timestamp_ms: block.block_timestamp,
+                    order,
+                    sim_value: None,
+                }
+            })
+            .collect();
+
+        trace!(
+            "Fetched synthetic transactions from CSV for block {}, batch {}, count: {}",
+            block.block_number,
+            batch_number,
+            orders.len()
+        );
+
+        Ok(orders)
+    }
+
+    fn clone_box(&self) -> Box<dyn DataSource> {
+        Box::new(self.clone())
+    }
+}
+
+fn transaction_to_order(
+    block: u64,
+    uuid_num: &mut u128,
+    tx: TransactionSignedEcRecovered,
+) -> Order {
+    let uuid_bytes = uuid_num.to_be_bytes();
+    let tx_with_blobs = TransactionSignedEcRecoveredWithBlobs::new_no_blobs(tx).unwrap();
+    let bundle = Bundle {
```

Comment:
Bundle creation is tricky (didn't have time to correct it :().
You should create it with a dummy hash/uuid (Default::default()) and then call hash_slow().